File: fast_mm_python/src/verify/VerifyMu.py
"""
Verification routine for mu bounds.
"""


def VerifyMu(q_, datapath):
    """
    Verify mu bound.
    
    Args:
        q_: q parameter value
        datapath: Path to the data file containing parameters
    """
    from ..control import InitExp, set_q
    from ..evaluation import Workspace, set_workspace
    from ..io import LoadParam
    from .GetFeasibility import GetFeasibility
    
    # Set global q
    set_q(q_)
    max_level = 3
    
    # Initialize experiment
    InitExp('__verify', 'mu', False, None)
    
    # Build workspace
    workspace = Workspace()
    workspace.Build(max_level)
    set_workspace(workspace)
    
    # Load parameters
    LoadParam(datapath, None, workspace.param_manager, workspace, None)
    
    # Get feasibility
    maxViol = GetFeasibility(workspace.param_manager, workspace, None)
    K = workspace.param_manager.cur_x[workspace.GetKPos()]
    
    print(f"mu <= {K:.8f} \t(MaxViolation: {maxViol:.6e})")
    
    if maxViol > 1.1e-6:
        print("[WARN] The last result seems wrong (the MaxViolation is too large).")
    elif maxViol > 1.1e-9:
        print("[WARN] The last result is not very accurate (MaxViolation > 1e-9).")

File: fast_mm_matlab/src/verify/VerifyMu.m
% This function reads data from a file and verify if omega(1, 1, K) <= 1 + 2 * K.
% (mu is defined as the solution of omega(1, 1, K) == 1 + 2 * K. It is not difficult to see that
%  once the above inequality holds, the K forms an upper bound of mu.)

function VerifyMu(q_, datapath)
  global workspace param_manager max_level q;
  q = q_;
  max_level = 3;
  InitExp('__verify', 'mu', false);
  workspace = Workspace();
  workspace.Build(max_level);
  LoadParam(datapath);

  maxViol = GetFeasibility();  % Get the max violation of all constraints.
  K = param_manager.cur_x(workspace.GetKPos());
  % omega = param_manager.cur_x(workspace.GetOmegaPos());

  fprintf("mu <= %.8f \t(MaxViolation: %e)\n", K, maxViol);
  if maxViol > 1.1e-6
    fprintf("[WARN] The last result seems wrong (the MaxViolation is too large).\n");
  elseif maxViol > 1.1e-9
    fprintf("[WARN] The last result is not very accurate (MaxViolation > 1e-9).\n");
  end
end
